README
Additional information and a guide can be found at
http://qmailwiki.inter7.com/index.php?title=Simscan

= Overview =
SimScan is a simplified scanner for qmail similar to qmail-scanner and qscand.
It uses clamav, trophie, and/or spamassassin.  It also supports attachment
blocking by extension.  Simscan is written entirely in C to ensure maximum
speed.  There are several options to allow simscan to scan per domain, and
reject spam mail.


= Requirements = 
# ripmime (If you plan on using attachment blocking)<br>
# qmail with qmail-queue patch<br>
# clamav (optional)<br>
# spamassassin (optional)<br>
# trophie (or sophie) (both optional)


= How it works =
* Simscan creates a temporary working directory. You can specify the base
working directory with the --enable-workdir=/path. The default location
of this base is the  /var/qmail/simscan directory. The temporary working
directory under this bass directory is named as
"unix time in seconds" . "microseconds" . "process id".

The email is read into a temporary file named "msg.unixtime.micro.pid.

* ripmime is called to break this file into separate mime parts.

* Optionally (--enable-attach) the mime message parts are checked against
the list of attachments to block.

Put the list of attachments in /var/qmail/control/ssattach or place the list
in the /var/qmail/control/simcontrol file for per-domain blocking.
This file gets read each time simscan is kicked off.

* clamdscan is called to check all the files in the working directory.
The return code of clamdscan is checked to see if it found a virus.
If it did, simscan exits back to qmail-smtpd with a permenent error,
causing the mail to be bounced back to the sender.

* Optionally (--enable-spam) spam assassin is called via spamc.
The email returned by spamc is checked X-Spam-Flag: header.  If this header
contains "YES", the spam is rejected.  You can reject instead on the calculated
hit count value by adding the --enable-spam-hits.  SimScan checks for both
"hits=" and "score=" so that both old and new versions of spamassassin are
supported.

* Optionally (--enable-per-domain) enable per domain clamav and
spamassassin processing. You will need to edit /var/qmail/control/simcontrol
to define what domains get what scans.  Further details on editing this file
are below.

Then run /var/qmail/bin/simscanmk to build the cdb files that simscan
uses. The changes will take effect immediately.

* Finally, if all of the above succeeds then the msg file is passed on to
qmail-queue. And the working directory and all the temporary files are deleted,
unless the SIMSCAN_DEBUG environment variable is set.

= Configuration Options =
Following is a list of all of the configure options and defaults:

  --enable-user=<user>              Change the user for simscan.
                                    Default: simscan.
  --enable-clamav=y|n               Turn on clamav scanning. default yes.
  --enable-clamdscan=PATH           Full path to clamdscan program.
  --enable-custom-smtp-reject=y|n   Return smtp reject message with virus name.
  --enable-per-domain=y|n           Turn on per domain based checking.
  --enable-attach=y|n               Turn on attachment scanning. default no.
  --enable-spam=y|n                 Turn on spam scanning. default no.
  --enable-spam-passthru=y|n        Pass spam email thru or reject
                                    Default: disable (reject)
  --enable-spamc-user=y|n           Set user option to spamc.
  --enable-spam-hits=number         Reject spam above this hit level.
                                    Default 10.0
  --enable-spamc=PATH               Full path to spamc program.
  --enable-spamc-args=ARGS          Arguments to pass to spamc.
  --enable-dropmsg=y|n              Drop message in case of virus/spam found.
                                    Don't return error to sender.
                                    Default: disable (return error)
  --enable-quarantinedir=DIR        Directory to keep spam and/or infected emails
                                    Default: disabled
  --enable-qmaildir=DIR             Base qmail directory /var/qmail.
  --enable-workdir=DIR              Directory to unpack emails
                                    Default: /var/qmail/simscan
  --enable-qmail-queue=PATH         Full path to qmail-queue program.
  --enable-trophie-path=PATH        Full path to the trophie binary.
  --enable-trophie-socket=PATH      Full path to the trophie socket.
  --enable-ripmime=PATH             Full path to ripmime program.
  --enable-received=y|n             Simscan should add a received line showing
                                    the version of all scanners that checked
                                    the message

These options are only needed when the received line option (--enabled-received=y)
and the corresponding scanners are enabled as well:

  --enable-spamassassin-path=PATH   Path to the spamassassin binary
  --enable-clamavdb-path=PATH       Directory where the clamav master.cvd and
                                    daily.cvd files are saved
  --enable-sigtool-path=PATH        Path to the sigtool binary


= Configuration Details =
Below are more detailed descriptions of each configuration option.

--enable-user=<user>
   This option defines the user that simscan will run as.  By default, the
   user is 'simscan'.

--enable-clamav=y|n
   This option turns clamav scanning on or off.  When enabled, an incoming
   email will be rejected if a virus is detected in the email.

--enable-clamdscan=PATH
   This option defines the path to the clamdscan binary.  This is the full
   path to the binary.
   Note : This option does nothing when clamav is not enabled.

--enable-custom-smtp-reject=y|n
   This option turns custom smtp reject messages on and off.  When enabled
   simscan will place the virus name in the reject message if a virus is
   detected.
   Note : The qmail-queue-custom-error.patch is needed to make this option
          work properly.  You can find this patch in the contrib directory.
   Note 2 : Enabling dropmsg disables this option

--enable-per-domain=y|n
   This option turns per-domain scanning on and off.  Per domain scanning
   allows the administrator to explicitly state what scanning occurs for
   what domain.  In addition, attachment scanning can be enabled or disabled
   for each domain.  Details about how to set up per-domain scanning are
   below.

--enable-attach=y|n
   This option turns on attachment scanning.  Attachment scanning will block
   all attachments specified in /var/qmail/control/ssattach.  Attachment
   scanning is disabled by default.

--enable-spam=y|n
   Ths option turns spam scanning on and off.  When enabled, simscan allows
   mail over a certain spam threshold to be rejected back to the sender.

--enable-spam-passthru=y|n
   This option turns spam passthru on and off.  When enabled, email
   identified as spam via the X-Spam-Status: header will be passed on to the
   user instead of rejected.
   Note : Enabling spam-hits effectively disables this option.

--enable-spamc-user=y|n
   This option turns per-user spamassassin on or off.  When enabled, the email
   address of the first rcpt to user is sent to spamassassin.  This allows
   spamassassin to use customized rules and settings for that email.

--enable-spam-hits=number
   This option specifies the number of hits a spam must receive to be rejected
   by simscan.  This option defaults to 10 hits.
   Note : This option disables spam passthru

--enable-spamc=PATH
   This option specifies the full path to the spamc binary.

--enable-spamc-args=ARGS
   This option defines the arguments to pass to spamc.  Be sure to place quotes
   around the options you define.

--enable-dropmsg=y|n
   This option causes messaged to be dropped when a virus/spam is found, rather
   than returning a 5xx error to the sender.  This option is disabled by default.
   Note : This option overrides the Custom SMTP Reject option
   Note 2 : If SPAM Passthru is enabled, SPAM will NOT be dropped unless 
            spam-hits is enabled

--enable-quarantinedir=DIR
   This option defined a directory to keep spam and/or infected emails.  This
   option is disabled by default.

--enable-qmaildir=DIR
   This option defines the location of qmail.

--enable-workdir=DIR
   This option defines the location of the working directory.  Note : The
   default directory is /var/qmail/simscan

--enable-qmail-queue=PATH
   This option defines the full path and name of the qmail-queue program.
   Incoming mail is passed to this program after being scanned by SimScan.

--enable-trophie-path=PATH
   This option defines the full path to the trophie binary. This option is only
   necessary if the received line option (--enable-received) is chosen.

--enable-trophie-socket=PATH
   This option defines the path to the trophie socket.  Defining this option
   enables trophie scanning.

--enable-ripmime=PATH
   This option defines the path to the ripmime program.  This program is used
   to rip apart emails into files.

--enable-received=y|n
   Simscan adds a Received line, showing the runtime, and a version string of
   the scanners that checked the message:
   Received: by simscan 1.0.6 ppid: 25399, pid: 25400, t: 4.7007s
         scanners: attach: 1.0.6 clamav: 0.80/m:27/d:556 trophie: 7.000-1011/218/74141 spam: 2.63

--enable-spamassassin-path=PATH
   This option defines the full path to the spamassassin binary.  This option
   is only necessary if the received line option (--enable-received) is chosen.

--enable-clamavdb-path=PATH
   This option defines the full path to clamav master.cvd and daily.cvd files.
   This option is only necessary if the received line option (--enable-received)
   is chosen.

--enable-sigtool-path=PATH
   This option defines the full path to the sigtool binary.  This option is only
   necessary if the received line option (--enable-received) is chosen.


= Attachment blocking option =
Attachments can be blocked. It is disabled by default.
If you use the per domain scanning as well, look in that section of the
documentation on how to enable the checking.
With only the attachment blocker and NO per domain option it works like
this:

Put the list of attachments in /var/qmail/control/ssattach.
List each attachment on it's own line. For example:
  [user@mailserver user]$ cat /var/qmail/control/ssattach
  .jpg
  .mp3
  .scr
  .bat

Then configure with:
  ./configure --enable-attach (with any other options you want)
  make
  make install-strip


= SpamAssassin options =
-------------------------------------------------
There are four different ways to configure simscan with spamassassin
* no spam processing
  This is the default. No spamassassin processing will be done.

* Reject email at smtp level if spamassassin considers it spam. This might reject false positives (good email that looks like spam)
  --enable-spam

* Reject email above a "score" of some number, like 15 and pass everything else through to the user.
  --enable-spam --enable-spam-hits=15

* Do not reject anything. Pass the spamassassin processed message through to the user
  --enable-spam --enable-spam-passthru

In addtion you can enable per user preference processing for email with just one recipient. Add the following option
  --enable-spamc-user

Look at the rc.spamd sample startup script for spamd options that work with vpopmail and per user preferences. Also look at the contrib directory for patches to spamassassin to make vpopmail/per user processing work.


= Trophie/Sophie option =
To enable the trophie virus scanner :

  ./configure --enable-trophie-socket=/path/to/the/socket \
    --enable-trophie-path=/path/to/trophie/binary

As sophie uses the same interface it may work as well,
but is not tested!


= How SMTP rejection works =
There are currently three options for handling SMTP rejection.  SMTP
rejection occurs when a virus is detected, or when the spam score is
over the spam-hits level.

The default rejection option is to reject with a standard 5XX reject
message.  This rejection message looks as follows :
  554 mail server permanently rejected message (#5.3.0)

If --enable-dropmsg is used, messages are dropped with no rejection
message.  The connection is simply closed without warning.

If --enable-custom-smtp-reject is used, messages are rejected with
a custom message.  You will need to apply the qmail-queue-custom-error.patch
patch located in the contrib directory in order to make this work.

For virus rejection, the message contains the name of the virus such as :

  Your email was rejected because it contains the Worm.Bagle.AU virus

For spam rejection, the message is more generic, merely stating that
the message was rejected because it was considered spam :


ChangeLog
01/23/2014 Johannes Weberhofer
        - Removed empty INSTALL file, added .gitattributes
        - renamed configure.in and fixed it to work with newer automake

01/22/2014 Johannes Weberhofer
        - Fixed issue in combination with a ripmime-problem as described
          in http://permalink.gmane.org/gmane.mail.qmail.simscan/3880
          (Thanks to Gustavo Castro)

30/05/2013 Johannes Weberhofer
         - Added GPL license version 2.0
         - Made several changes to allow building on with strict compiler 
           settings
         - Removed .cvsignore file and added .gitignore file
         - Do no longer check for existing clamav-database when set via 
           --enable-clamavdb-path
         - added missing permissions S_IRUSR | S_IWUSR to open()

12/12/09 Jason Frisvold
         - Make simscan aware of clamav cld files (Thanks Bob Maple)
         - Fix ripmime working directory (Thanks Gladiston Justini)

03/21/08 Jason Frisvold
         - Fix --enable-dpsam-args (Thanks Jean-Daniel Bonnetot)
         - Prevent double virus scanning (Thanks Gerard Earley)
         - Fix retrieval of SpamHits and ReqHits (Thanks anonymous)
         - Add SpamAssassin's ReqHits to log line (Thanks anonymous)
         - Add SPAM PASS log line (Thanks anonymous)

11/12/07 Jason Frisvold
         - Improvement patch (Tom Collins)
         - - Clamdscan scans the message file instead of the message file and
         -   attachment
         - - Don't pass messages over 250k to spamassassin
         - - Spam scanning before virus scanning
         - - Lowercase FROM and TO addresses
         - - write spam_fd before calling is_spam()
         - - fix str_rstr() call to require complete match

10/30/07 Jason Frisvold
         - simscan directory umask fix (John Simpson)
         - band-aid simscanmk NFS timing issue on solaris - Tom Collins
         - strsep bug in Solaris - Tom Collins
         - set exit code on errors - Jeremy Kister
         - fix compile warning issues with long versus int - Jeremy Kister
         - add pid to debug messages - Jeremy Kister
         - RFC2822 Received field compliance - Tullio Andreatta

10/29/07 Jason Frisvold
         - Support daily and main .inc and .cvd files (John Simpson)
         - Fixed a few compiler warnings (Jason Frisvold)
         - simcontrol lookups should be lowercase (Jason Frisvold)
         - lowercase all strings in the simcontrol.cdb file (Jason Frisvold)

10/29/07 Ken Jones
         - release version 1.4.0 with p0f code moved to qmail-smtpd

03/11/07 Ken Jones
         - for dspam only authorized users can send email to spam@ and nospam@,
           includes support for custom smtp error message

05/18/06 George Breahna
         - first version of DSPAM support

         R
         - add u_char casting to isupper and tolower

01/07/06 Ken Jones
         - envoke ripmime with --disable-qmail-bounce

01/07/06 Ken Jones
         - reject blank envelopes with proper error

10/05/05 Ken Jones
         - found an IsSpam=1.0 instead of IsSpam==1
            for the spam passthru setting

10/04/05 Ken Jones
         - virus drop message was not exiting at the right time.
             Found by Russ Lists
         - use per domain hit count on spam log message if set.
         - if read from standard in fails the temp file was not
             closed so the temporary work directory would not
             be removed. 

09/30/05 Ken Jones
         - put 's' at the end of processing time and
            move to after status.
         - add RELAYCLIENT log line. suggested by Sim
         - replace --with-pcre with two new ones
            --with-pcre-lib --with-pcre-include
         - update configure regex section to report the
            two new --with-pcre lines
         - A value of 5 was being checked instead of
            the SPAM_HITS value. Found by Jeremy Kister.
         - Change SpamHits > to SpamHits >=
           

09/28/05 Ken Jones
         - update logging documentation on qmailwiki.org
         - add VIRUS DROPPED logging message
         - clearer --enable-spam-auth-user config message
         - put #ifdef ENABLE_SPAM around the spam processing
            section in per_domain_lookup() function 
            Reported by Gerad Earley
         - Change received header printf to %d for compiler warnings
         - add SPAM DROPPED log message
         - add processing time field to logging
         - add spam score and required hits to logging

09/27/05 Ken Jones
         - New global IsSpam flag
         - rewrite is_spam() to assume one line per invocation
         - rewrite check_spam() to use FILE and fgets to call
             is_spam with one line per invocation
         - Use IsSpam flag instead of over loaded SpamHits=1.0
             to determine if email is spam. This also returns
             a valid spam hit count in custom smtp error message
         - move definition of qmail_queue outside per domain ifdef
         - new configure option --enable-spam-auth-user default to
           not scan authenticated users with spamassasin
         - update copyright year
         - add standarized reporting line for attachment and regex
            rejects

09/27/05 Bruno Negrao
         - in per domain processing an attachment list should override
           the global attachment list 

02/08/05 Tom Collins
         - simscan.c do not spam scan for RELAYCLIENT users
         - complier warnings for getpid(),getppid()

02/06/05 Itamar Reis Peixoto
         - rc.spamd sample init script change to remove clamd with spamd
         - rc.qmail same change to this init script sample

02/02/05 Ken Jones
         - 1.1 release 

01/26/05 Alex Pleiner   
         - clean up for addr parse bug and per domain pass through debug

01/21/05 Flavio Curti
         - spam passthru change in how it returns a value

         Daniel Maglione
		 - spam passthru log changes
        

11/30/04 Ken Jones / Tren Blackburn

simscanmk.c
/*
 * $Id: simscanmk.c,v 1.5 2009/12/13 04:39:38 xen0phage Exp $
 * Copyright (C) 2004-2005 Inter7 Internet Technologies, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 */
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <sys/param.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/select.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <arpa/nameser.h>
#include <fcntl.h>
#include <resolv.h>
#include <errno.h>
#include <pthread.h>
#include <ctype.h>
#include "config.h"

#include "cdb/cdb_make.h"
#include "cdb/cdb.h"
#include "cdb/buffer.h"

static struct cdb_make c;


#define MAX_KEY 40
#define MAX_DATA 40
#define MAX_LINE 4000
#define MAX_HOST_SIZE 200
#define MAX_IPS 200
#define MAX_HOSTS 200

#define BUILD_CLAM 1
#define BUILD_SPAM 2

void lowerit(char *input);
int make_cdb();
void usage();
void get_options(int argc,char **argv);
char ClearFile[200];
char CdbFile[200];
char CdbTmpFile[200];

#define TOKENS ":\n\t\r #"
#ifdef ENABLE_RECEIVED
int buildversions=0;
char CdbVersFile[200];
int make_version_cdb();
#endif

int main( int argc, char **argv)
{
  get_options(argc,argv);
#ifdef ENABLE_RECEIVED
  if (buildversions){
    int xcode = make_version_cdb();
    if (xcode != 0) exit(xcode);
  } else {
#endif
  int xcode = make_cdb();
  if (xcode != 0) exit(xcode);
#ifdef ENABLE_RECEIVED
  }
#endif
  exit(0);
}

void lowerit(char *input)
{
  while(*input!=0) {
    if ( isupper((u_char)*input) ) {
      *input = (u_char)tolower((u_char)*input);
    }
    ++input;
  }
}

#ifdef ENABLE_RECEIVED

int add_cdb_key (struct cdb_make *cdb, char *key, char *data){
  uint32 h;
  char *tmpptr;
  if (cdb_make_addbegin(cdb,strlen(key),strlen(data)) == -1) {
    printf("error on cdb_make_addbegin)\n");
    return(-1);
  } 
  h = CDB_HASHSTART;
  
  for(tmpptr=key;*tmpptr!=0;++tmpptr) {
    if (buffer_PUTC(&cdb->b,*tmpptr) == -1) {
      printf("error in buffer_PUTC\n");
      return(-1);
    }
    h = cdb_hashadd(h,*tmpptr);
  }

  for(tmpptr=data;*tmpptr!=0;++tmpptr) {
    if (buffer_PUTC(&cdb->b,*tmpptr) == -1) {
      printf("error in buffer_PUTC\n");
      return(-1);
    }
  }
  if (cdb_make_addend(cdb,strlen(key),strlen(data),h) == -1) {
    printf("error in cdb_make_addend\n");
    return(-1);
  }
  return(0);
}

int make_version_cdb() {
  int fdout;
  char key[MAX_KEY];
  char data[MAX_DATA];
  pid_t pid;
  int pin[2],rmstat,r,f;
  char input[MAX_LINE];
  char dbpath[MAX_LINE];
  char *pos;
#if ENABLE_SPAM==1 || ENABLE_TROPHIE==1
  int fnd_vsvers;
#endif
#if ENABLE_TROPHIE==1
  int fnd_patvers;
#endif

  if ( (fdout = open(CdbTmpFile, O_CREAT | O_TRUNC | O_WRONLY, S_IRUSR | S_IWUSR )) < 0) {
    printf("error on open tmp file\n");
    return(-1);
  }
  
  if (cdb_make_start(&c,fdout) == -1) {
    printf("error on cdb_make_start\n");
    return(-1);
  } 

  /* now add a check for every enabled scanner, and add the scanner to the cdb */
#ifdef ENABLE_ATTACH
  memset(data,'\0',MAX_DATA);
  strncpy(key,RCVD_ATTACH_KEY,MAX_KEY);
  strncpy(data,VERSION,MAX_DATA);
  add_cdb_key(&c,key,data);
#endif
#ifdef ENABLE_REGEX
  memset(data,'\0',MAX_DATA);
  strncpy(key,RCVD_REGEX_KEY,MAX_KEY);
  strncpy(data,VERSION,MAX_DATA);
  add_cdb_key(&c,key,data);
#endif
#ifdef ENABLE_TROPHIE
  memset(data,'\0',MAX_DATA);
  data[0]='\0';
  fnd_vsvers=0;
  fnd_patvers=0;
  strncpy(key,RCVD_TROPHIE_KEY,MAX_KEY);
  if (pipe(pin)){
    printf("error opening pipe for trophie\n");
  }
  pid=vfork();
  if (pid==0){
      /* in the child */
      close(pin[0]);
      dup2(pin[1],2); /* stderr goes to the pipe */
      execl(TROPHIEBINARY,TROPHIEBINARY,"-v",NULL);
      printf("error running trophie\n");
      _exit(-1); /* we should never get here! */
  } else if (pid){
    /* in the parent */
    close(pin[1]);
    while((r=read(pin[0],input,MAX_LINE))){
      /* we are looking for those two lines:
          Initializing    : VSAPI version 6.150-1001
          Initializing    : Pattern version 218 (pattern number 51417)
        and we want 6.150-1001/218/51417 in the string at the end */

      input[r]='\0';
      if ( (pos=strstr(input,"version ")) && (!fnd_vsvers || !fnd_patvers)){
        if (!fnd_vsvers && !fnd_patvers){
          /* this line is the vsapi version */
          strncpy(data,pos+8,strlen(pos)-9);
          strcat(data,"/");
          fnd_vsvers=1;
        } else if (!fnd_patvers){
          /* this line is the pattern version */
          for(f=0;*(pos+8+f)!=' ' && *(pos+8+f)!='\0';f++);
          strncpy(data+strlen(data),pos+8,f);
          if ( (pos=strstr(input,"number ")) ) {
            strcat(data,"/");
            for(f=0;*(pos+7+f)!=')' && *(pos+7+f)!='\0';f++);
            strncpy(data+strlen(data),pos+7,f);
          }
          fnd_patvers=1;
        }
      }
    }
    waitpid(pid,&rmstat,0);
    add_cdb_key(&c,key,data);
  } else {
    printf("error forking for trophie\n");
  }
  close(pin[0]); close(pin[1]);
#endif
#ifdef ENABLE_SPAM
  memset(data,'\0',MAX_DATA);
  fnd_vsvers=0;
  strncpy(key,RCVD_SPAM_KEY,MAX_KEY);
  if (pipe(pin)){
    printf("error opening pipe for spamassassin\n");
  }
  pid=vfork();
  if (pid==0){
      /* in the child */
      close(pin[0]);
      dup2(pin[1],1); /* stdout goes to the pipe */
      execl(SPAMASSASSINPATH,SPAMASSASSINPATH,"-V",NULL);
      printf("error running spamassassin\n");
      _exit(-1); /* we should never get here! */
  } else if (pid){
    /* in the parent */
    close(pin[1]);
    while((r=read(pin[0],input,MAX_LINE))){
      /* we are looking for this line:
          SpamAssassin version 2.63
          and have 2.63 as version
        */
      input[r]='\0';
      if ( (pos=strstr(input,"version ")) && !fnd_vsvers ){
        /* this line is the sa version */
        for(f=0;*(pos+8+f)!='\n' && *(pos+8+f)!='\0';f++);
        strncpy(data,pos+8,f);
        fnd_vsvers=1;
      }
    }
    waitpid(pid,&rmstat,0);
    add_cdb_key(&c,key,data);
  } else {
    printf("error forking for trophie\n");
  }
  close(pin[0]); close(pin[1]);
#endif
#ifdef ENABLE_CLAMAV
  memset(data,'\0',MAX_DATA);
  strncpy(key,RCVD_CLAM_KEY,MAX_KEY);
  if (pipe(pin)){
    printf("error opening pipe for sigtool\n");
  }
  pid=vfork();
  if (pid==0){
      /* in the child */
      close(pin[0]);
      dup2(pin[1],1); /* stdout goes to the pipe */
      execl(CLAMDSCAN,CLAMDSCAN,"--stdout","-V",NULL);
      printf("error running clamdscan\n");
      _exit(-1); /* we should never get here! */
  } else if (pid){
    /* in the parent */
    close(pin[1]);
    while((r=read(pin[0],input,MAX_LINE))){
      /* we are looking for this line:
         Version: 27
        */
      input[r]='\0';
      if ( (pos=strstr(input,"ClamAV "))){
        /* this line is the db version */
        for(f=0;*(pos+7+f)!='/' && *(pos+7+f)!='\0';f++);
        strncat(data,pos+7,f);
        strcat(data,"/");
      }
    }
    waitpid(pid,&rmstat,0);
    close(pin[0]); close(pin[1]);
  }
  strncpy(dbpath,CLAMAVDBPATH,MAX_LINE);
  strncat(dbpath,"/main.inc/main.info",(MAX_LINE-sizeof(CLAMAVDBPATH)-1));
  if(access(dbpath,F_OK)) {
    strncpy(dbpath,CLAMAVDBPATH,MAX_LINE);
    strncat(dbpath,"/main.cvd",(MAX_LINE-sizeof(CLAMAVDBPATH)-1));
  }
  if(access(dbpath,F_OK)) {
    strncpy(dbpath,CLAMAVDBPATH,MAX_LINE);
    strncat(dbpath,"/main.cld",(MAX_LINE-sizeof(CLAMAVDBPATH)-1));
  }
  strcat(data,"m:");
  if (pipe(pin)){
    printf("error opening pipe for sigtool\n");
  }
  pid=vfork();
  if (pid==0){
      /* in the child */
      close(pin[0]);
      dup2(pin[1],1); /* stdout goes to the pipe */
      execl(SIGTOOLPATH,SIGTOOLPATH,"--stdout","-i",dbpath,NULL);
      printf("error running sigtool\n");
      _exit(-1); /* we should never get here! */
  } else if (pid){
    /* in the parent */
    close(pin[1]);
    while((r=read(pin[0],input,MAX_LINE))){
      /* we are looking for this line:
         Version: 27
        */
      input[r]='\0';
      if ( (pos=strstr(input,"Version: "))){
        /* this line is the db version */
        for(f=0;*(pos+9+f)!='\n' && *(pos+9+f)!='\0';f++);
        strncat(data,pos+9,f);
      }
    }
    waitpid(pid,&rmstat,0);
    close(pin[0]); close(pin[1]);
    strncpy(dbpath,CLAMAVDBPATH,MAX_LINE);
    strncat(dbpath,"/daily.inc/daily.info",(MAX_LINE-sizeof(CLAMAVDBPATH)-1));
    if(access(dbpath,F_OK)) {
      strncpy(dbpath,CLAMAVDBPATH,MAX_LINE);
      strncat(dbpath,"/daily.cvd",(MAX_LINE-sizeof(CLAMAVDBPATH)-1));
    }
    if(access(dbpath,F_OK)) {
      strncpy(dbpath,CLAMAVDBPATH,MAX_LINE);
      strncat(dbpath,"/daily.cld",(MAX_LINE-sizeof(CLAMAVDBPATH)-1));
    }
    if (pipe(pin)){
     printf("error opening pipe for sigtool\n");
    }
    pid=vfork();
    if (pid==0){
      /* in the child */
      close(pin[0]);
      dup2(pin[1],1); /* stdout goes to the pipe */
      execl(SIGTOOLPATH,SIGTOOLPATH,"--stdout","-i",dbpath,NULL);
      printf("error running sigtool\n");
      _exit(-1); /* we should never get here! */
    } else if (pid){
      /* in the parent */
      close(pin[1]);
      while((r=read(pin[0],input,MAX_LINE))){
        /* we are looking for this line:
           Version: 27
          */
        input[r]='\0';
        if ( (pos=strstr(input,"Version: "))){
          /* this line is the db version */
          for(f=0;*(pos+9+f)!='\n' && *(pos+9+f)!='\0';f++);
          strcat(data,"/d:");
          strncat(data,pos+9,f);
        }
      }
      waitpid(pid,&rmstat,0);
      add_cdb_key(&c,key,data);
    }
  } else {
    printf("error forking for trophie\n");
  }
  close(pin[0]); close(pin[1]);
#endif
#ifdef ENABLE_AVIRA
  memset(data,'\0',MAX_DATA);
  strncpy(key,RCVD_AVIRA_KEY,MAX_KEY);
  if (pipe(pin)){
    printf("error opening pipe for Avira antivir\n");
  }
  pid=vfork();
  if (pid==0){
      /* in the child */
      close(pin[0]);
      dup2(pin[1],1); /* stdout goes to the pipe */
      execl(AVIRABINARY,AVIRABINARY,"--version",NULL);
      printf("error running Avira antivir\n");
      _exit(-1); /* we should never get here! */
  } else if (pid){
    /* in the parent */
    close(pin[1]);
    while((r=read(pin[0],input,MAX_LINE))){
      /* we are looking for this line:
         vdf version:      X.XX
        */
      input[r]='\0';
      
      if ( (pos=strstr(input,"product version: "))){
        /* this line is the db version */
        for(f=0;*(pos+18+f)!='\n' && *(pos+18+f)!='\0';f++);
        strncat(data,pos+18,f);
      }
      strcat(data, "/e:");
      if ( (pos=strstr(input,"engine version: "))){
        /* this line is the db version */
        for(f=0;*(pos+18+f)!='\n' && *(pos+18+f)!='\0';f++);
        strncat(data,pos+18,f);
      }
      strcat(data, "/v:");
      if ( (pos=strstr(input,"vdf version: "))){
        /* this line is the db version */
        for(f=0;*(pos+18+f)!='\n' && *(pos+18+f)!='\0';f++);
        strncat(data,pos+18,f);
      }
    }
    waitpid(pid,&rmstat,0);
    add_cdb_key(&c,key,data);
  } else {
    printf("error forking for Avira antivir\n");
  }
  close(pin[0]); close(pin[1]);
#endif
  if (cdb_make_finish(&c) == -1) {
    printf("error in cdb_make_finish\n"); 
    return(-1);
  }

  close(fdout);

  if (rename(CdbTmpFile, CdbVersFile)==-1) {
    printf("error: could not rename %s to %s\n", CdbTmpFile, CdbFile);
    return(-1);
  }
  chmod(CdbVersFile, 0644);
  printf("simscan versions cdb file built. %s\n", CdbVersFile);
  return(0);
}
#endif

int make_cdb()
{
 FILE *fs;
 int fdout;
 char *key;
 char *data;
 char *tmpptr;
 static char input[MAX_LINE];
 uint32 h;

  sleep(0); /* some NFS timing crazyness on solaris jk 20061108 */

  if ( (fs = fopen(ClearFile,"r")) == NULL) {
    printf("Not building simcontrol.cdb file. No %s/simcontrol text file\n",
      CONTROLDIR);
    return(-1);
  }
 
  if ( (fdout = open(CdbTmpFile, O_CREAT | O_TRUNC | O_WRONLY, S_IRUSR | S_IWUSR )) < 0) {
     printf("error on open tmp file\n");
    return(-1);
  }

  if (cdb_make_start(&c,fdout) == -1) {
    printf("error on cdb_make_start\n");
    return(-1);
  } 

  while(fgets(input,sizeof(input), fs)!=NULL) { 

    if ( input[0] == ':' ) {
      key = "";
      data = strtok(&input[1],"\r\n");
    } else {
      key = strtok(input,TOKENS);
      if ( key == NULL ) continue;
      data = strtok(NULL,"\r\n");
      // Lowercase all characters
      lowerit(key);
      lowerit(data);
    }

    if ( data == NULL ) data="";

    /*snprintf(key,sizeof(key),"%s", data);*/

    if (cdb_make_addbegin(&c,strlen(key),strlen(data)) == -1) {
      printf("error on cdb_make_addbegin)\n");
      return(-1);
    } 
    h = CDB_HASHSTART;
    
    for(tmpptr=key;*tmpptr!=0;++tmpptr) {
      if (buffer_PUTC(&c.b,*tmpptr) == -1) {
        printf("error in buffer_PUTC\n");
        return(-1);
      }
      h = cdb_hashadd(h,*tmpptr);
    }

    for(tmpptr=data;*tmpptr!=0;++tmpptr) {
      if (buffer_PUTC(&c.b,*tmpptr) == -1) {
        printf("error in buffer_PUTC\n");
        return(-1);
      }
    }
    if (cdb_make_addend(&c,strlen(key),strlen(data),h) == -1) {
      printf("error in cdb_make_addend\n");
      return(-1);
    }
  }
  if (cdb_make_finish(&c) == -1) {
    printf("error in cdb_make_finish\n"); 
    return(-1);
  }

  /*fprintf(fsout"\n");*/
  fclose(fs);
  close(fdout);

  if (rename(CdbTmpFile, CdbFile)==-1) {
    printf("error: could not rename %s to %s\n", CdbTmpFile, CdbFile);
    return(-1);
  }
  chmod(CdbFile, 0644);
  printf("simscan cdb file built. %s/simcontrol.cdb\n", CONTROLDIR);
  return(0);
}

void get_options(int argc,char **argv)
{
 int c;
 int errflag;
 extern char *optarg;
 extern int optind;

  snprintf(ClearFile, sizeof(ClearFile), "%s/simcontrol", CONTROLDIR);
  snprintf(CdbFile, sizeof(CdbFile), "%s/simcontrol.cdb", CONTROLDIR);

  snprintf(CdbTmpFile, sizeof(CdbTmpFile), "%s/ss.cdb.tmp.%d", 
    CONTROLDIR, getpid());

  errflag = 0;
  while( !errflag && (c=getopt(argc,argv,"g")) != -1 ) {
    switch(c) {
#ifdef ENABLE_RECEIVED
      case 'g':
        snprintf(CdbVersFile, sizeof(CdbFile), 
          "%s/simversions.cdb", CONTROLDIR);
        buildversions=1;
        break;
#endif
      default:
        errflag = 1;
        break;
    }
  }
  if ( errflag > 0 ) usage();
}

void usage()
{
#ifdef ENABLE_RECEIVED
  printf("usage: simscanmk [-g]\n");
  printf("       no params: builds %s/simcontrol.cdb from %s/simcontrol\n", 
    CONTROLDIR, CONTROLDIR);
  printf("       -g: builds %s/simversions.cdb from scanners\n", CONTROLDIR);
#else
  printf("usage: no options\n");
  printf("builds %s/simcontrol.cdb from %s/simcontrol\n", 
    CONTROLDIR, CONTROLDIR);
#endif
  exit(1);
}
